README.rst
============================
Nelson-Siegel-Svensson Model
============================


.. image:: https://img.shields.io/pypi/v/nelson_siegel_svensson.svg
        :target: https://pypi.python.org/pypi/nelson_siegel_svensson

.. image:: https://github.com/luphord/nelson_siegel_svensson/actions/workflows/python-package.yml/badge.svg
        :target: https://github.com/luphord/nelson_siegel_svensson/actions

.. image:: https://readthedocs.org/projects/nelson-siegel-svensson/badge/?version=latest
        :target: https://nelson-siegel-svensson.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status




Implementation of the Nelson-Siegel-Svensson interest rate curve model in Python.

.. code-block:: python

        from nelson_siegel_svensson import NelsonSiegelSvenssonCurve
        import numpy as np
        from matplotlib.pyplot import plot

        y = NelsonSiegelSvenssonCurve(0.028, -0.03, -0.04, -0.015, 1.1, 4.0)
        t = np.linspace(0, 20, 100)
        plot(t, y(t))

.. image:: docs/_static/an_example_nelson-siegel-svensson-curve.png

* Free software: MIT license
* Python 3.7 or later supported
* Documentation: https://nelson-siegel-svensson.readthedocs.io.


Features
--------

* Python implementation of the Nelson-Siegel curve (three factors)
* Python implementation of the Nelson-Siegel-Svensson curve (four factors)
* Methods for zero and forward rates (as vectorized functions of time points)
* Methods for the factors (as vectorized function of time points)
* Calibration based on ordinary least squares (OLS) for betas and nonlinear optimization for taus
* Simple command line interface (CLI) for evaluating, calibrating and plotting curves


Calibration
-----------

In order to calibrate a curve to given data you can use the `calibrate_ns_ols` and
`calibrate_nss_ols` functions in the `calibrate` module:

.. code-block:: python

        import numpy as np
        from nelson_siegel_svensson.calibrate import calibrate_ns_ols

        t = np.array([0.0, 0.5, 1.0, 2.0, 3.0, 4.0, 5.0, 10.0, 15.0, 20.0, 25.0, 30.0])
        y = np.array([0.01, 0.011, 0.013, 0.016, 0.019, 0.021, 0.026, 0.03, 0.035, 0.037, 0.038, 0.04])

        curve, status = calibrate_ns_ols(t, y, tau0=1.0)  # starting value of 1.0 for the optimization of tau
        assert status.success
        print(curve)

which gives the following output:

.. code-block:: python

        NelsonSiegelCurve(beta0=0.04201739383636799, beta1=-0.031829031569430594, beta2=-0.026797319779108236, tau=1.7170972656534174)

.. image:: docs/_static/calibrated_nelson-siegel-curve.png


Command Line interface
----------------------

`nelson_siegel_svensson` provides basic functionality using a command line interface (CLI):

.. code-block:: console

        Usage: nelson_siegel_svensson [OPTIONS] COMMAND [ARGS]...

        Commandline interface for nelson_siegel_svensson.

        Options:
        --help  Show this message and exit.

        Commands:
        calibrate  Calibrate a curve to the given data points.
        evaluate   Evaluate a curve at given points.
        plot       Plot a curve at given points.


In order to calibrate a curve to given data points on the command line, try

.. code-block:: console

        nelson_siegel_svensson calibrate -t '[0.0, 0.5, 1.0, 2.0, 3.0, 4.0, 5.0, 10.0, 15.0, 20.0, 25.0, 30.0]' -y '[0.01, 0.011, 0.013, 0.016, 0.019, 0.021, 0.026, 0.03, 0.035, 0.037, 0.038, 0.04]' --nelson-siegel --initial-tau1 1.0

which gives

.. code-block:: console

        {"beta0": 0.042017393764903765, "beta1": -0.03182903146166806, "beta2": -0.026797320316066128, "tau": 1.717097232403383}

This curve can then be evaluated on the command line using

.. code-block:: console

        nelson_siegel_svensson evaluate -c '{"beta0": 0.042017393764903765, "beta1": -0.03182903146166806, "beta2": -0.026797320316066128, "tau": 1.717097232403383}' -t '[0, 1, 2, 3]'

resulting in

.. code-block:: console

        [0.010188362303235707, 0.012547870204470839, 0.01574855552855885, 0.01897955804146046]

And finally, the curve can be plotted with

.. code-block:: console

        nelson_siegel_svensson plot -o cli_plot_example.png -c '{"beta0": 0.042017393764903765, "beta1": -0.03182903146166806, "beta2": -0.026797320316066128, "tau": 1.717097232403383}'

.. image:: docs/_static/cli_plot_example.png

Note that the quoting in the above commands prevents `bash` from evalutating the JSON-based parameters. Depending on your shell, you may require a different quoting mechanism.

Credits
-------

Main developer is luphord_.

.. _luphord: https://github.com/luphord

This package was prepared with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.5.0 (2022-11-13)
------------------

* Drop support for Python 3.7
* Add support Python 3.9 and 3.10
* Reformat code with black
* Upgrade dependencies
* Upgrade development status to beta

0.4.3 (2022-03-15)
------------------

* Migrate to GitHub Actions

0.4.2 (2020-02-04)
------------------

* Support Python 3.8
* Upgrade flake8 (test requirement) as prior version did not support Python 3.8

0.4.1 (2019-11-14)
------------------

* Added return type annotations for core modules
* Added some example notebooks

0.4.0 (2019-07-08)
------------------

* Simple command line interface (CLI) supporting curve evaluation, calibration and plotting
* Added more documentation

0.3.0 (2019-03-17)
------------------

* Added type annotations

0.2.0 (2019-02-20)
------------------

* Ordinary least squares based calibration of Nelson-Siegel-Svensson
* Ordinary least squares based calibration of Nelson-Siegel
* A little bit of usage documentation

0.1.0 (2019-02-13)
------------------

* First release on PyPI.


tests/__init__.py
# -*- coding: utf-8 -*-

'''Unit test package for nelson_siegel_svensson.'''


tests/test_nelson_siegel_svensson.py
# -*- coding: utf-8 -*-

import unittest
import os
import json
from dataclasses import asdict

import numpy as np
import click
from click.testing import CliRunner

from nelson_siegel_svensson import cli, NelsonSiegelCurve, NelsonSiegelSvenssonCurve


class TestNelson_siegel_svensson(unittest.TestCase):
    """Tests for `nelson_siegel_svensson` CLI."""

    def setUp(self):
        self.y1 = NelsonSiegelCurve(0.017, -0.023, 0.24, 2.2)
        self.y2 = NelsonSiegelSvenssonCurve(0.017, -0.023, 0.24, 0.1, 2.2, 3.1)
        self.t = [0.0, 0.5, 1.0, 2.0, 3.0, 4.0, 5.0, 10.0, 15.0, 20.0, 25.0, 30.0]
        self.y = [
            0.01,
            0.011,
            0.013,
            0.016,
            0.019,
            0.021,
            0.026,
            0.03,
            0.035,
            0.037,
            0.038,
            0.04,
        ]
        self.runner = CliRunner()

    def test_command_line_interface(self):
        """Test the CLI."""
        result = self.runner.invoke(cli.cli_main)
        self.assertEqual(0, result.exit_code)
        help_result = self.runner.invoke(cli.cli_main, ["--help"])
        self.assertEqual(0, help_result.exit_code)
        self.assertIn("--help  Show this message and exit.", help_result.output)

    def test_cli_evaluate(self):
        """Test evaluate CLI."""
        param = [
            "evaluate",
            "-c",
            '{"beta0": 0.017, "beta1": -0.023, "beta2": 0.24, "tau": 2}',
            "-t",
            "[1,2,3]",
        ]
        result = self.runner.invoke(cli.cli_main, param)
        self.assertEqual(0, result.exit_code)
        self.assertIn("0.0758359", result.output)

    def test_cli_calibrate(self):
        """Test calibrate CLI."""
        param = ["calibrate", "-t", json.dumps(self.t), "-y", json.dumps(self.y)]
        result = self.runner.invoke(cli.cli_main, param)
        self.assertEqual(0, result.exit_code)
        self.assertIn("0.0451", result.output)
        first_output = result.output
        result = self.runner.invoke(cli.cli_main, param + ["--nelson-siegel-svensson"])
        self.assertEqual(0, result.exit_code)
        self.assertEqual(first_output, result.output)
        result = self.runner.invoke(cli.cli_main, param + ["--nelson-siegel"])
        self.assertEqual(0, result.exit_code)
        self.assertIn("0.0425", result.output)
        result = self.runner.invoke(
            cli.cli_main, param + ["--nelson-siegel", "--initial-tau1", "1.234"]
        )
        self.assertEqual(0, result.exit_code)
        self.assertIn("0.04179", result.output)

    def test_cli_plot(self):
        """Test plot CLI."""
        fname = "output.png"
        param = [
            "plot",
            "-o",
            fname,
            "-c",
            '{"beta0": 0.017, "beta1": -0.023, "beta2": 0.24, "tau": 2}',
        ]
        with self.runner.isolated_filesystem():
            result = self.runner.invoke(cli.cli_main, param)
            self.assertEqual(0, result.exit_code)
            self.assertTrue(os.path.exists(fname), fname + " missing")
            result = self.runner.invoke(cli.cli_main, param + ["-f", "10", "-t", "20"])
            self.assertEqual(0, result.exit_code)
            self.assertTrue(os.path.exists(fname), fname + " missing")

    def test_curve_parameters(self):
        """Test curve parameter."""
        param = cli.Curve()
        self.assertRaises(
            click.BadParameter, param.convert, value="", param=None, ctx=None
        )
        self.assertRaises(
            click.BadParameter, param.convert, value="{}", param=None, ctx=None
        )
        missing_tau = '{"beta0": 0.017, "beta1": -0.023, "beta2": 0.24}'
        self.assertRaises(
            click.BadParameter, param.convert, value=missing_tau, param=None, ctx=None
        )
        self.assertEqual(
            self.y1, param.convert(json.dumps(asdict(self.y1)), None, None)
        )
        self.assertEqual(
            self.y2, param.convert(json.dumps(asdict(self.y2)), None, None)
        )

    def test_float_array_parameters(self):
        """Test float array parameter."""
        param = cli.FloatArray()
        self.assertRaises(
            click.BadParameter, param.convert, value="", param=None, ctx=None
        )
        self.assertRaises(
            click.BadParameter, param.convert, value='{"a": 1}', param=None, ctx=None
        )
        self.assertRaises(
            click.BadParameter, param.convert, value='["a"]', param=None, ctx=None
        )
        self.assertEqual(np.array([1.0]), param.convert("[1.0]", None, None))
        self.assertEqual(np.array([1.0]), param.convert("[1]", None, None))
        self.assertEqual([], param.convert("[]", None, None).tolist())
        self.assertTrue(
            (
                np.array([1.0, 2.0, 3.0]) == param.convert("[1,   2,3.0]", None, None)
            ).all()
        )


tests/test_nelson_siegel_calibration.py
# -*- coding: utf-8 -*-

import unittest

import numpy as np
from scipy.optimize import minimize

from nelson_siegel_svensson import NelsonSiegelCurve
from nelson_siegel_svensson.calibrate import (
    betas_ns_ols,
    errorfn_ns_ols,
    calibrate_ns_ols,
)


class TestNelsonSiegelCurveCalibration(unittest.TestCase):
    """Tests for Nelson-Siegel curve calibration."""

    def setUp(self):
        self.y = NelsonSiegelCurve(0.017, -0.023, 0.24, 2.2)

    def test_nelson_siegel_betas_recovery(self):
        """Test recovery of betas using ordinary least squares
        (given fixed tau).
        """
        t = np.linspace(0, 30)
        y_target = self.y(t)
        curve, lstsq_res = betas_ns_ols(self.y.tau, t, y_target)
        self.assertAlmostEqual(self.y.beta0, curve.beta0, places=12)
        self.assertAlmostEqual(self.y.beta1, curve.beta1, places=12)
        self.assertAlmostEqual(self.y.beta2, curve.beta2, places=12)

    def test_nelson_siegel_ols_errorfn_for_correct_tau(self):
        """Test ols based error function for correct tau."""
        t = np.linspace(0, 30)
        y_target = self.y(t)
        error = errorfn_ns_ols(self.y.tau, t, y_target)
        self.assertAlmostEqual(0.0, error, places=12)
        error2 = errorfn_ns_ols(self.y.tau * 1.1, t, y_target)
        self.assertNotAlmostEqual(0.0, error2)

    def test_nelson_siegel_ols_calibration(self):
        """Test ols based calibration of Nelson-Siegel model."""
        t = np.linspace(0, 30)
        y_target = self.y(t)
        y_hat, opt_res = calibrate_ns_ols(t, y_target, tau0=self.y.tau)
        places = 12
        self.assertAlmostEqual(self.y.beta0, y_hat.beta0, places=places)
        self.assertAlmostEqual(self.y.beta1, y_hat.beta1, places=places)
        self.assertAlmostEqual(self.y.beta2, y_hat.beta2, places=places)
        self.assertAlmostEqual(self.y.tau, y_hat.tau, places=places)
        # The following does not work in general, but in this case the
        # error function (of tau) is well-behaving
        places_beta = 3
        places_tau = 2
        y_hat, opt_res = calibrate_ns_ols(t, y_target)  # default start value
        self.assertAlmostEqual(self.y.beta0, y_hat.beta0, places=places_beta)
        self.assertAlmostEqual(self.y.beta1, y_hat.beta1, places=places_beta)
        self.assertAlmostEqual(self.y.beta2, y_hat.beta2, places=places_beta)
        self.assertAlmostEqual(self.y.tau, y_hat.tau, places=places_tau)
        y_hat, opt_res = calibrate_ns_ols(t, y_target, tau0=0.5)
        self.assertAlmostEqual(self.y.beta0, y_hat.beta0, places=places_beta)
        self.assertAlmostEqual(self.y.beta1, y_hat.beta1, places=places_beta)
        self.assertAlmostEqual(self.y.beta2, y_hat.beta2, places=places_beta)
        self.assertAlmostEqual(self.y.tau, y_hat.tau, places=places_tau)
        y_hat, opt_res = calibrate_ns_ols(t, y_target, tau0=5)
        self.assertAlmostEqual(self.y.beta0, y_hat.beta0, places=places_beta)
        self.assertAlmostEqual(self.y.beta1, y_hat.beta1, places=places_beta)
        self.assertAlmostEqual(self.y.beta2, y_hat.beta2, places=places_beta)
        self.assertAlmostEqual(self.y.tau, y_hat.tau, places=places_tau)
        y_hat, opt_res = calibrate_ns_ols(t, y_target, tau0=10)
        self.assertAlmostEqual(self.y.beta0, y_hat.beta0, places=places_beta)
        self.assertAlmostEqual(self.y.beta1, y_hat.beta1, places=places_beta)
        self.assertAlmostEqual(self.y.beta2, y_hat.beta2, places=places_beta)
        self.assertAlmostEqual(self.y.tau, y_hat.tau, places=places_tau)
        y_hat, opt_res = calibrate_ns_ols(t, y_target, tau0=20)
        self.assertAlmostEqual(self.y.beta0, y_hat.beta0, places=places_beta)
        self.assertAlmostEqual(self.y.beta1, y_hat.beta1, places=places_beta)
        self.assertAlmostEqual(self.y.beta2, y_hat.beta2, places=places_beta)
        self.assertAlmostEqual(self.y.tau, y_hat.tau, places=places_tau)

    def test_nelson_siegel_ols_constrained_calibration(self):
        """Test ols based calibration with constrained optimization of tau
        of Nelson-Siegel model."""
        t = np.linspace(0, 30)
        y_target = self.y(t)
        tau0_grid = (0.08, 0.1, 0.2, 0.5, 1, 2, 3, 4, 5)
        places_tau = 2
        # unconstrained
        for tau0 in tau0_grid:
            opt_res = minimize(errorfn_ns_ols, x0=tau0, args=(t, y_target))
            self.assertAlmostEqual(self.y.tau, opt_res.x[0], places=places_tau)
        # constrained
        for tau0 in tau0_grid:
            opt_res = minimize(
                errorfn_ns_ols, x0=tau0, args=(t, y_target), bounds=((0.08, None),)
            )
            self.assertAlmostEqual(self.y.tau, opt_res.x[0], places=places_tau)


nelson_siegel_svensson/ns.py
# -*- coding: utf-8 -*-

"""Implementation of a Nelson-Siegel interest rate curve model.
See `NelsonSiegelCurve` class for details.
"""

from numbers import Real
from dataclasses import dataclass
from typing import Union, Tuple

import numpy as np
from numpy import exp

EPS = np.finfo(float).eps


@dataclass
class NelsonSiegelCurve:
    """Implementation of a Nelson-Siegel interest rate curve model.
    This curve can be interpreted as a factor model with three
    factors (including a constant).
    """

    beta0: float
    beta1: float
    beta2: float
    tau: float

    def factors(
        self, T: Union[float, np.ndarray]
    ) -> Union[Tuple[float, float], Tuple[np.ndarray, np.ndarray]]:
        """Factor loadings for time(s) T, excluding constant."""
        tau = self.tau
        if isinstance(T, Real) and T <= 0:
            return 1, 0
        elif isinstance(T, np.ndarray):
            zero_idx = T <= 0
            T[zero_idx] = EPS  # avoid warnings in calculations
        exp_tt0 = exp(-T / tau)
        factor1 = (1 - exp_tt0) / (T / tau)
        factor2 = factor1 - exp_tt0
        if isinstance(T, np.ndarray):
            T[zero_idx] = 0
            factor1[zero_idx] = 1
            factor2[zero_idx] = 0
        return factor1, factor2

    def factor_matrix(self, T: Union[float, np.ndarray]) -> Union[float, np.ndarray]:
        """Factor loadings for time(s) T as matrix columns,
        including constant column (=1.0).
        """
        factor1, factor2 = self.factors(T)
        constant: Union[float, np.ndarray] = (
            np.ones(T.size) if isinstance(T, np.ndarray) else 1
        )
        return np.stack([constant, factor1, factor2]).transpose()

    def zero(self, T: Union[float, np.ndarray]) -> Union[float, np.ndarray]:
        """Zero rate(s) of this curve at time(s) T."""
        factor1, factor2 = self.factors(T)
        return self.beta0 + self.beta1 * factor1 + self.beta2 * factor2

    def __call__(self, T: Union[float, np.ndarray]) -> Union[float, np.ndarray]:
        """Zero rate(s) of this curve at time(s) T."""
        return self.zero(T)

    def forward(self, T: Union[float, np.ndarray]) -> Union[float, np.ndarray]:
        """Instantaneous forward rate(s) of this curve at time(s) T."""
        exp_tt0 = exp(-T / self.tau)
        return self.beta0 + self.beta1 * exp_tt0 + self.beta2 * exp_tt0 * T / self.tau


nelson_siegel_svensson/nss.py
# -*- coding: utf-8 -*-

"""Implementation of a Nelson-Siegel-Svensson interest rate curve model.
See `NelsonSiegelSvenssonCurve` class for details.
"""

from numbers import Real
from dataclasses import dataclass
from typing import Union, Tuple

import numpy as np
from numpy import exp

EPS = np.finfo(float).eps


@dataclass
class NelsonSiegelSvenssonCurve:
    """Implementation of a Nelson-Siegel-Svensson interest rate curve model.
    This curve can be interpreted as a factor model with four
    factors (including a constant).
    """

    beta0: float
    beta1: float
    beta2: float
    beta3: float
    tau1: float
    tau2: float

    def factors(
        self, T: Union[float, np.ndarray]
    ) -> Union[Tuple[float, float, float], Tuple[np.ndarray, np.ndarray, np.ndarray]]:
        """Factor loadings for time(s) T, excluding constant."""
        tau1 = self.tau1
        tau2 = self.tau2
        if isinstance(T, Real) and T <= 0:
            return 1, 0, 0
        elif isinstance(T, np.ndarray):
            zero_idx = T <= 0
            T[zero_idx] = EPS  # avoid warnings in calculations
        exp_tt1 = exp(-T / tau1)
        exp_tt2 = exp(-T / tau2)
        factor1 = (1 - exp_tt1) / (T / tau1)
        factor2 = factor1 - exp_tt1
        factor3 = (1 - exp_tt2) / (T / tau2) - exp_tt2
        if isinstance(T, np.ndarray):
            T[zero_idx] = 0
            factor1[zero_idx] = 1
            factor2[zero_idx] = 0
            factor3[zero_idx] = 0
        return factor1, factor2, factor3

    def factor_matrix(self, T: Union[float, np.ndarray]) -> Union[float, np.ndarray]:
        """Factor loadings for time(s) T as matrix columns,
        including constant column (=1.0).
        """
        factor1, factor2, factor3 = self.factors(T)
        constant: Union[float, np.ndarray] = (
            np.ones(T.size) if isinstance(T, np.ndarray) else 1.0
        )
        return np.stack([constant, factor1, factor2, factor3]).transpose()

    def zero(self, T: Union[float, np.ndarray]) -> Union[float, np.ndarray]:
        """Zero rate(s) of this curve at time(s) T."""
        beta0 = self.beta0
        beta1 = self.beta1
        beta2 = self.beta2
        beta3 = self.beta3
        factor1, factor2, factor3 = self.factors(T)
        res = beta0 + beta1 * factor1 + beta2 * factor2 + beta3 * factor3
        return res

    def __call__(self, T: Union[float, np.ndarray]) -> Union[float, np.ndarray]:
        """Zero rate(s) of this curve at time(s) T."""
        return self.zero(T)

    def forward(self, T: Union[float, np.ndarray]) -> Union[float, np.ndarray]:
        """Instantaneous forward rate(s) of this curve at time(s) T."""
        exp_tt0 = exp(-T / self.tau1)
        exp_tt1 = exp(-T / self.tau2)
        return (
            self.beta0
            + self.beta1 * exp_tt0
            + self.beta2 * exp_tt0 * T / self.tau1
            + self.beta3 * exp_tt1 * T / self.tau2
        )


nelson_siegel_svensson/calibrate.py
# -*- coding: utf-8 -*-

"""Calibration methods for Nelson-Siegel(-Svensson) Models.
See `calibrate_ns_ols` and `calibrate_nss_ols` for ordinary least squares
(OLS) based methods.
"""

from typing import Tuple, Any

import numpy as np
from numpy.linalg import lstsq
from scipy.optimize import minimize

from .ns import NelsonSiegelCurve
from .nss import NelsonSiegelSvenssonCurve


def _assert_same_shape(t: np.ndarray, y: np.ndarray) -> None:
    assert t.shape == y.shape, "Mismatching shapes of time and values"


def betas_ns_ols(
    tau: float, t: np.ndarray, y: np.ndarray
) -> Tuple[NelsonSiegelCurve, Any]:
    """Calculate the best-fitting beta-values given tau
    for time-value pairs t and y and return a corresponding
    Nelson-Siegel curve instance.
    """
    _assert_same_shape(t, y)
    curve = NelsonSiegelCurve(0, 0, 0, tau)
    factors = curve.factor_matrix(t)
    lstsq_res = lstsq(factors, y, rcond=None)
    beta = lstsq_res[0]
    return NelsonSiegelCurve(beta[0], beta[1], beta[2], tau), lstsq_res


def errorfn_ns_ols(tau: float, t: np.ndarray, y: np.ndarray) -> float:
    """Sum of squares error function for a Nelson-Siegel model and
    time-value pairs t and y. All betas are obtained by ordinary
    least squares given tau.
    """
    _assert_same_shape(t, y)
    curve, lstsq_res = betas_ns_ols(tau, t, y)
    return np.sum((curve(t) - y) ** 2)


def calibrate_ns_ols(
    t: np.ndarray, y: np.ndarray, tau0: float = 2.0
) -> Tuple[NelsonSiegelCurve, Any]:
    """Calibrate a Nelson-Siegel curve to time-value pairs
    t and y, by optimizing tau and chosing all betas
    using ordinary least squares.
    """
    _assert_same_shape(t, y)
    opt_res = minimize(errorfn_ns_ols, x0=tau0, args=(t, y))
    curve, lstsq_res = betas_ns_ols(opt_res.x[0], t, y)
    return curve, opt_res


def empirical_factors(
    y_3m: float, y_2y: float, y_10y: float
) -> Tuple[float, float, float]:
    """Calculate the empirical factors according to
    Diebold and Li (2006)."""
    return y_10y, y_10y - y_3m, 2 * y_2y - y_3m - y_10y


def betas_nss_ols(
    tau: Tuple[float, float], t: np.ndarray, y: np.ndarray
) -> Tuple[NelsonSiegelSvenssonCurve, Any]:
    """Calculate the best-fitting beta-values given tau (= array of tau1
    and tau2) for time-value pairs t and y and return a corresponding
    Nelson-Siegel-Svensson curve instance.
    """
    _assert_same_shape(t, y)
    curve = NelsonSiegelSvenssonCurve(0, 0, 0, 0, tau[0], tau[1])
    factors = curve.factor_matrix(t)
    lstsq_res = lstsq(factors, y, rcond=None)
    beta = lstsq_res[0]
    return (
        NelsonSiegelSvenssonCurve(beta[0], beta[1], beta[2], beta[3], tau[0], tau[1]),
        lstsq_res,
    )


def errorfn_nss_ols(tau: Tuple[float, float], t: np.ndarray, y: np.ndarray) -> float:
    """Sum of squares error function for a Nelson-Siegel-Svensson
    model and time-value pairs t and y. All betas are obtained
    by ordinary least squares given tau (= array of tau1
    and tau2).
    """
    _assert_same_shape(t, y)
    curve, lstsq_res = betas_nss_ols(tau, t, y)
    return np.sum((curve(t) - y) ** 2)


def calibrate_nss_ols(
    t: np.ndarray, y: np.ndarray, tau0: Tuple[float, float] = (2.0, 5.0)
) -> Tuple[NelsonSiegelSvenssonCurve, Any]:
    """Calibrate a Nelson-Siegel-Svensson curve to time-value
    pairs t and y, by optimizing tau1 and tau2 and chosing
    all betas using ordinary least squares. This method does
    not work well regarding the recovery of true parameters.
    """
    _assert_same_shape(t, y)
    opt_res = minimize(errorfn_nss_ols, x0=np.array(tau0), args=(t, y))
    curve, lstsq_res = betas_nss_ols(opt_res.x, t, y)
    return curve, opt_res


nelson_siegel_svensson/cli.py
# -*- coding: utf-8 -*-

"""Console script for nelson_siegel_svensson."""
import sys
import click
import json

import numpy as np
from matplotlib import pyplot as plt

from .ns import NelsonSiegelCurve
from .nss import NelsonSiegelSvenssonCurve
from .calibrate import calibrate_ns_ols, calibrate_nss_ols


class Curve(click.ParamType):
    """Parameter type representing a curve (either Nelson-Siegel
    or Nelson-Siegel-Svensson)"""

    name = "curve"

    def convert(self, value, param, ctx):
        try:
            decoded = json.loads(value)
            if "beta3" in decoded:
                return NelsonSiegelSvenssonCurve(**decoded)
            else:
                return NelsonSiegelCurve(**decoded)
        except Exception as e:
            self.fail("{} is not a valid curve: {}".format(value, e), param, ctx)


class FloatArray(click.ParamType):
    """Parameter type representing an array of floats"""

    name = "floats"

    def convert(self, value, param, ctx):
        try:
            decoded = json.loads(value)
            return np.array([float(f) for f in decoded])
        except Exception as e:
            self.fail(
                "{} is not a valid array of floats: {}".format(value, e), param, ctx
            )


@click.group(name="nelson_siegel_svensson")
def cli_main(args=None):
    """Commandline interface for nelson_siegel_svensson."""
    return 0


@click.command(name="calibrate")
@click.option(
    "-t", "--times", type=FloatArray(), required=True, help="time points as JSON array."
)
@click.option(
    "-y",
    "--values",
    type=FloatArray(),
    required=True,
    help="values corresponding to time points as JSON array.",
)
@click.option(
    "--nelson-siegel-svensson/--nelson-siegel",
    default=True,
    help="whether to calibrate a Nelson-Siegel-Svensson (4 factor)"
    + " or a Nelson-Siegel (3 factor) curve. Defaults to"
    + " the former.",
)
@click.option(
    "--initial-tau1",
    type=click.FLOAT,
    default=2.0,
    show_default=True,
    help="Initial value of tau1 (or tau for Nelson-Siegel) " + "for optimization.",
)
@click.option(
    "--initial-tau2",
    type=click.FLOAT,
    default=5.0,
    show_default=True,
    help="Initial value of tau2 (ignored by Nelson-Siegel) " + "for optimization.",
)
def cli_calibrate(times, values, nelson_siegel_svensson, initial_tau1, initial_tau2):
    """Calibrate a curve to the given data points."""
    if nelson_siegel_svensson:
        curve, status = calibrate_nss_ols(times, values, (initial_tau1, initial_tau2))
    else:
        curve, status = calibrate_ns_ols(times, values, initial_tau1)
    assert status.success
    click.echo(json.dumps(vars(curve)))


@click.command(name="evaluate")
@click.option(
    "-c",
    "--curve",
    type=Curve(),
    required=True,
    help="Parameters for curve as JSON object.",
)
@click.option(
    "-t",
    "--times",
    type=FloatArray(),
    required=True,
    help="Evaluation times as JSON array.",
)
def cli_evaluate(curve, times):
    """Evaluate a curve at given points."""
    click.echo(json.dumps(curve(times).tolist()))


@click.command(name="plot")
@click.option(
    "-c",
    "--curves",
    type=Curve(),
    required=True,
    multiple=True,
    help="Parameters for curve as JSON object; " + "accepts multiple curves.",
)
@click.option(
    "-o",
    "--output",
    type=click.Path(file_okay=True, dir_okay=False, writable=True, resolve_path=True),
    required=True,
    help="Output file for the plot.",
)
@click.option(
    "-f",
    "--from-time",
    type=click.FLOAT,
    default=0.0,
    show_default=True,
    help="Left time point of the plot.",
)
@click.option(
    "-t",
    "--to-time",
    type=click.FLOAT,
    default=30.0,
    show_default=True,
    help="Right time point of the plot.",
)
def cli_plot(curves, output, from_time, to_time):
    """Plot a curve at given points."""
    fig, ax = plt.subplots(nrows=1, ncols=1)
    t = np.linspace(from_time, to_time, num=100)
    for curve in curves:
        y = curve(t)
        ax.plot(t, y)
    fig.savefig(output)
    plt.close(fig)


cli_main.add_command(cli_calibrate)
cli_main.add_command(cli_evaluate)
cli_main.add_command(cli_plot)


if __name__ == "__main__":
    sys.exit(cli_main())  # pragma: no cover
